src/AutoSchedule.cpp

```diff
@@ -0,0 +1,2893 @@
+            children.erase(prod_group);
+            for (auto &f : children) {
+                set<FStage> &cons = f.second;
+                auto iter = cons.find(prod_group);
+                if (iter != cons.end()) {
+                    cons.erase(iter);
+                    // For a function with multiple stages, all the stages will
+                    // be in the same group and the consumers of the function
+                    // only depend on the last stage. Therefore, when the
+                    // producer group has multiple stages, parents of the
+                    // producers should point to the consumers of the last
+                    // stage of the producer.
+                    cons.insert(prod_group_children.begin(), prod_group_children.end());
+                }
+            }
+        }
+
+        Cost post_merge = get_pipeline_cost();
+
+        disp_pipeline_costs();
+
+        internal_assert((pre_merge.arith + pre_merge.memory) >=
+                        (post_merge.arith + post_merge.memory));
+    }
+}
+
+DimBounds Partitioner::get_bounds(const FStage &s) {
+    Definition def = get_stage_definition(s.func, s.stage_num);
+    DimBounds bounds;
+
+    const vector<string> &args = s.func.args();
+    for (size_t d = 0; d < args.size(); d++) {
+        bounds[args[d]] = get_element(pipeline_bounds, s.func.name())[d];
+    }
+
+    return get_stage_bounds(s.func, s.stage_num, bounds);
+}
+
+DimBounds Partitioner::get_bounds_from_tile_sizes(const FStage &s,
+                                                  const map<string, int> &tile_sizes) {
+    Definition def = get_stage_definition(s.func, s.stage_num);
+    map<string, Interval> bounds;
+
+    const map<string, Interval> &def_bounds = get_bounds(s);
+    const vector<Dim> &dims = def.schedule().dims();
+
+    for (int d = 0; d < (int)dims.size() - 1; d++) {
+        string var = dims[d].var;
+        const Interval &bound = get_element(def_bounds, var);
+        const auto &iter = tile_sizes.find(var);
+        if (iter != tile_sizes.end()) {
+            int size = iter->second;
+            // Check if the bounds allow for tiling with the given tile size,
+            // i.e. ensure at least 2 tiles
+            int64_t extent = get_extent(bound);
+            if (extent >= 2 * size) {
+                // TODO: Maybe shift this to the center of the pipeline bound
+                bounds[var] = Interval(0, size - 1);
+            } else {
+                // If the dimension is too small, do not tile it and set the
+                // extent of the bounds to that of the dimension estimate
+                bounds[var] = bound;
+            }
+        } else {
+            bounds[var] = bound;
+        }
+    }
+
+    return bounds;
+}
+
+Partitioner::GroupAnalysis Partitioner::analyze_group(const Group &g, bool show_analysis) {
+    // Get the definition corresponding to the group output
+    Definition def = get_st
```

Comment:
I agree generating a readable schedule from a func is a much general functionality that is useful. A representation of the schedule which is not tied to the variable names and can be reapplied is good to have.